addons/crm/models/crm_lead.py

```diff
@@ -1241,6 +1258,73 @@ def _handle_salesmen_assignment(self, user_ids=False, team_id=False):
     # CLASSIFICATION TOOLS
     # --------------------------------------------------
 
+    def _get_potential_lead_duplicates(self, include_lost=False, limit=20):
+        """ Returns the leads that could be considered as "duplicate".
+        A lead will be considered as duplicate if:
+            - The lead is linked to the given partner
+            - The lead has a similar email address
+        If the provided email address has a common domain name (e.g: google.com, yahoo.com, ...),
+        the function will retrieve the leads having the same email address.
+        Otherwise, the function will retrieve the leads having the same email domain.
+        :param boolean include_lost: If true, the leads considered as lost will be included in the results
+        :param boolean limit: The function will return at most "limit + 1" potential duplicates
+        """
+        k = 1
+        limit = 20
```

Comment:
1. Not really: When we have an initial domain like `A OR B` and we remove `B`, we remove the results sitting in the region `B AND NOT A`. Here, we can still select an element having the property B if the element has other properties that we consider as relevant.
2. It is an arbitrary limit. We could maybe let the user set it.

Here, I started from a broad filter and I gradually add/remove clauses to be more restrictive. Of course, we could do the opposite. We can actually keep these two strategies and choose the best one on case-by-case basis. What I really like with my approach is that we can swap from one strategy to another by iterating from `0 to 1` or from `1 to 0`.

It is also possible to add more granularity by adding more levels. Of course, we should avoid having many levels because, in the worst case scenario, we will have to execute n queries sequentially which can lead to high loading time.

I can maybe add a better documentation to explain the algorithm.